### ISSUE TITLE
ICE: Module validation failed: Invalid bitcast

### ISSUE THREAD

**Issue by Sija** [crystal-lang/crystal#8198]
Code below:

```cr
class Object
  def to_foo(io); end

  def to_foo
    String.build &->to_foo(IO)
  end
end

Time.local.to_foo # for structs works fine apparently
:foobar.to_foo  # ... but it crashes having simple `Value`s as implicit `self`
```

generates following ICE:

```
Module validation failed: Invalid bitcast
  %6 = bitcast i32 %self to i8*
 (Exception)
  from Crystal::CodeGenVisitor#finish:Nil
  from Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil)
  from Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result
  from Crystal::Command#run_command<Bool>:Nil
  from Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil)
  from main
```

See https://carc.in/#/r/7le1

**Comment by asterite** [comment]
Related to #7577: you currently can't get a proc of a primitive (number, symbol, etc.) method

**Comment by HertzDevil** [comment]
[Cannot reproduce since 0.36](https://godbolt.org/z/dTe8d4hYj). This is most likely fixed by #9824
